src/binary/non_blocking/binary_buffer.rs

```diff
@@ -0,0 +1,808 @@
+        }
+
+        incomplete_data_error("a VarUInt", self.total_consumed() + encoded_size_in_bytes)
+    }
+
+    /// Reads a `VarInt` encoding primitive from the beginning of the buffer. If it is successful,
+    /// returns an `Ok(_)` containing its [VarInt] representation and consumes the source bytes.
+    ///
+    /// See: https://amzn.github.io/ion-docs/docs/binary.html#varuint-and-varint-fields
+    pub fn read_var_int(&mut self) -> IonResult<VarInt> {
+        const BITS_PER_ENCODED_BYTE: usize = 7;
+        const STORAGE_SIZE_IN_BITS: usize = mem::size_of::<i64>() * 8;
+        const MAX_ENCODED_SIZE_IN_BYTES: usize = STORAGE_SIZE_IN_BITS / BITS_PER_ENCODED_BYTE;
+
+        const LOWER_6_BITMASK: u8 = 0b0011_1111;
+        const LOWER_7_BITMASK: u8 = 0b0111_1111;
+        const HIGHEST_BIT_VALUE: u8 = 0b1000_0000;
+
+        const BITS_PER_BYTE: usize = 8;
+        const BITS_PER_U64: usize = mem::size_of::<u64>() * BITS_PER_BYTE;
+
+        // Unlike VarUInt's encoding, the first byte in a VarInt is a special case because
+        // bit #6 (0-indexed, from the right) indicates whether the value is positive (0) or
+        // negative (1).
+
+        if self.is_empty() {
+            return incomplete_data_error("a VarInt", self.total_consumed());
+        }
+        let first_byte: u8 = self.peek_next_byte().unwrap();
+        let no_more_bytes: bool = first_byte >= 0b1000_0000; // If the first bit is 1, we're done.
+        let is_positive: bool = (first_byte & 0b0100_0000) == 0;
+        let sign: i64 = if is_positive { 1 } else { -1 };
+        let mut magnitude = (first_byte & 0b0011_1111) as i64;
+
+        if no_more_bytes {
+            self.consume(1);
+            return Ok(VarInt::new(magnitude * sign, !is_positive, 1));
+        }
+
+        let mut encoded_size_in_bytes = 1;
+        // Whether we found the terminating byte in this buffer.
+        let mut terminated = false;
+
+        for byte in self.bytes()[1..].iter().copied() {
+            let lower_seven = (0b0111_1111 & byte) as i64;
+            magnitude <<= 7;
+            magnitude |= lower_seven;
+            encoded_size_in_bytes += 1;
+            if byte >= 0b1000_0000 {
+                terminated = true;
+                break;
+            }
+        }
+
+        if !terminated {
+            // panic!("incomplete (unterm) varint @ pos {}: {:x?}", self.total_consumed(), self.bytes());
+            return incomplete_data_error(
+                "a VarInt",
+                self.total_consumed() + encoded_size_in_bytes,
+            );
+        }
+
+        if encoded_size_in_bytes > MAX_ENCODED_SIZE_IN_BYTES {
+            return decoding_error(format!(
+                "Found a {}-byte VarInt. Max supported size is {} bytes.",
+                encoded_size_in_bytes, MAX_ENCODED_SIZE_IN_BYTES
+            ));
+        }
+
+        self.consume(encoded_size_in_bytes);
+        Ok(VarInt::new(
+            magnitude * sign,
+            !is_positive,
```

Comment:
Suggestion: Everywhere we use `!is_positive` with `VarInt::new`, why not just use `is_negative` in declaration above and use it as is. Only reason I suggest this is because I had to go back and look at the definition of `VarInt::new` to make sure the above statement is correct.

Reply:
Done!